gmso/atomtyping/isomorph.py

```diff
@@ -0,0 +1,59 @@
+"""TopologyGraph Functions that identify molecules from isomorphism."""
+from collections import deque
+
+import networkx as nx
+
+
+def partition_isomorphic_topology_graphs(graph):
+    """Return a dictionary of subgraphs that are partitioned by isomorphism.
+
+     Parameters
+     ----------
+     graph: foyer.topology_graph.TopologyGraph
+         The networkx subclassed TopologyGraph with data identifying the nodes
+         and edges that make up a topology atom and bonds structure
+
+     Returns
+     -------
+     isomorphic_elements: dict
+         The keys are unique disconnected graphs, and the values are all
+         identical subgraphs in the graph
+
+    Notes
+    -----
+    See https://github.com/networkx/networkx/blob/main/networkx/algorithms/isomorphism/isomorphvf2.py
+    from the networkx documentation about identifying isomorphic components
+    """
+
+    subgraphs_gen = (graph.subgraph(c) for c in nx.connected_components(graph))
+    subgraphs_list = list(subgraphs_gen)
+
+    graph_queue = deque(subgraphs_list)
+    graph_of_interest = graph_queue.popleft()
+    isomorphic_elements = {
+        graph_of_interest: [],
+    }
+
+    last_element_popped = None
+    count = 0
+    first_mismatch = None
+    while graph_queue:
+        if graph_queue[0] == first_mismatch:
+            count = 0
+            graph_of_interest = graph_queue.popleft()
+            isomorphic_elements[graph_of_interest] = []
+        if graph_queue:
+            graph = graph_queue.popleft()
+            matcher = nx.algorithms.isomorphism.GraphMatcher(
+                graph, graph_of_interest
```

Comment:
I think we need to add a node_match function (based on the element of the node) IIRC.

Reply:
What exactly would a node_match function do? I'm not entirely familiar, is it something passed to the GraphMatcher to do the node matching?

Reply:
just tested this offline, I think the current use of GraphMatcher only concern about the shape of the graph but not about the nodes involved. So this could cause issue with molecules with same shape but different element would still be recognized as isomorphic (so lumped into the same group).

Reply:
```python3 
ethanol = mb.load("CCO", smiles=True)
ethanol.name = "Ethanol"

eth1 = mb.clone(ethanol)
eth2 = mb.clone(ethanol)

eth1_top = mb_convert(eth1)
eth1_top.sites[0].name = "K"
eth1_top.sites[0].element = element_by_symbol("K")
eth2_top = mb_convert(eth2)
```
Note below I am using the `get_modified_topology_graph` from the example notebook.
```python3
eth1_grp = get_modified_topology_graph(eth1_top)
eth2_grp = get_modified_topology_graph(eth2_top)
matcher = nx.isomorphism.GraphMatcher(eth1_grp, eth2_grp)
```
Then 
`matcher.is_isomorphic()` will still return `True`

Reply:
on a different note, I think we can use this `GraphMatcher` for the issue with the assumed site order in a molecule (or subtopology) (also this https://github.com/mosdef-hub/foyer/issues/504). Will have to sacrifice some performance for checking but the number of checks performed will still be fewer that going the completely splitting the topology, so should still be a bit faster.

Reply:
Okay this makes more sense to me now, thanks @daico007. I was a little hazy on the exact implementation of GraphMatcher. I think having a few different options for matching makes sense, which trade of increasing performance for increasing matching precision. This reminds me a bit of the filters used in PR #649, which can be used to specify how you want to filter unique parameter types. I think it makes sense to at least try a basic one that checks at least for elements in the correct positions, and allow that to be a built in option.